contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,546 @@
+    function startUpgrade() external {
+        _startUpgrade();
+    }
+
+    function finishUpgrade() external {
+        _finishUpgrade();
+    }
+
+    /// Perform basic checks to revert the entire upgrade if something gone wrong
+    function assertUpgradeIsFinishedCorrectly() external view {
+        _assertUpgradeIsFinishedCorrectly();
+    }
+
+    function revertIfUpgradeNotFinished() external view {
+        if (!isUpgradeFinished) {
+            revert UpgradeIsNotFinished();
+        }
+    }
+
+    function _startUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        if (isUpgradeStarted) revert CanOnlyStartOnce();
+        isUpgradeStarted = true;
+
+        _locator.proxy__upgradeTo(_locatorImplementation);
+
+        _assertCorrectInitialState();
+
+        // Upgrade proxy implementation
+        _upgradeProxyImplementations();
+
+        // Need to have the implementations already attached at this moment
+        _assertCorrectNonAdminRoleHolders();
+
+        _withdrawalVault().initialize();
+
+        _initializeWithdrawalQueue();
+
+        _initializeAccountingOracle();
+
+        _initializeValidatorsExitBus();
+
+        _migrateLidoOracleCommitteeMembers();
+
+        _initializeStakingRouter();
+
+        _migrateDSMGuardians();
+    }
+
+    function _assertCorrectInitialState() internal view {
+        if (ILidoOracle(address(_legacyOracle())).getVersion() != 3) revert LidoOracleMustNotBeUpgradedToLegacyYet();
+
+        _assertAdminsOfProxies(address(this));
+
+        _assertInitialDummyProxyImplementations();
+
+        // Check roles of non-proxy contracts (can do without binding implementations)
+        _assertSingleOZRoleHolder(_hashConsensusForAccountingOracle, DEFAULT_ADMIN_ROLE, address(this));
+        _assertSingleOZRoleHolder(_hashConsensusForValidatorsExitBusOracle, DEFAULT_ADMIN_ROLE, address(this));
+        _assertSingleOZRoleHolder(_burner(), DEFAULT_ADMIN_ROLE, address(this));
+        if (_depositSecurityModule().getOwner() != address(this)) revert WrongDsmOwner();
+    }
+
+    function _upgradeProxyImplementations() internal {
+        _accountingOracle().proxy__upgradeTo(_accountingOracleImplementation);
+        _validatorsExitBusOracle().proxy__upgradeTo(_validatorsExitBusOracleImplementation);
+        _stakingRouter().proxy__upgradeTo(_stakingRouterImplementation);
+        _withdrawalVault().proxy__upgradeTo(_withdrawalVaultImplementation);
+        _withdrawalQueue().proxy__upgradeTo(_withdrawalQueueImplementation);
+    }
+
+    function _assertAdminsOfProxies(address admin) internal view {
+        _assertProxyAdmin(_accountingOracle(), admin);
+        _assertProxyAdmin(_locator, admin);
+        _assertProxyAdmin(_stakingRouter(), admin);
+        _assertProxyAdmin(_validatorsExitBusOracle(), admin);
+        _assertProxyAdmin(_withdrawalQueue(), admin);
+        _assertProxyAdmin(_withdrawalVault(), admin);
```

Comment:
this one is different (already deployed, have `voting` as an admin)
maybe shouldn't be in this particular list